src/vm.rs

```diff
@@ -419,6 +424,60 @@ fn matches_literal(s: &str, ix: usize, end: usize, literal: &str) -> bool {
     end <= s.len() && &s.as_bytes()[ix..end] == literal.as_bytes()
 }
 
+#[inline]
+fn matches_literal_casei(s: &str, ix: usize, end: usize, literal: &str) -> bool {
+    if end > s.len() {
+        return false;
+    }
+    if matches_literal(s, ix, end, literal) {
+        return true;
+    }
+    if s.is_char_boundary(ix) && s.is_char_boundary(end) && s.is_ascii() {
+        return s[ix..end].eq_ignore_ascii_case(literal);
+    }
+
+    // text captured and being backreferenced is not ascii, so we utilize regex-automata's case insensitive matching
+
+    let chars = literal.chars();
+
+    use alloc::boxed::Box;
+    use regex_syntax::ast::*;
+
+    let span = Span::splat(Position::new(0, 0, 0));
+    let mut literals = Vec::with_capacity(literal.len());
+    for c in chars {
+        let ast_literal = Ast::Literal(Box::new(Literal {
+            span,
+            kind: LiteralKind::Verbatim,
+            c: c,
+        }));
+        literals.push(ast_literal);
+    }
+    let ast_flags = Ast::Group(Box::new(Group {
+        span,
+        kind: GroupKind::NonCapturing(Flags {
+            span,
+            items: vec![FlagsItem {
+                span,
+                kind: FlagsItemKind::Flag(Flag::CaseInsensitive),
+            }],
+        }),
+        ast: Box::new(Ast::Concat(Box::new(Concat {
+            span,
+            asts: literals,
+        }))),
+    }));
+
+    let mut translator = regex_syntax::hir::translate::Translator::new();
+    let hir = translator.translate(literal, &ast_flags).unwrap();
```

Comment:
Looks like we can simplify this a bit, like this:

```rust
    let span = Span::splat(Position::new(0, 0, 0));
    let literals = literal
        .chars()
        .map(|c| {
            Ast::literal(Literal {
                span,
                kind: LiteralKind::Verbatim,
                c: c,
            })
        })
        .collect();
    let ast = Ast::concat(Concat {
        span,
        asts: literals,
    });

    let mut translator = regex_syntax::hir::translate::TranslatorBuilder::new()
        .case_insensitive(true)
        .build();
    let hir = translator.translate(literal, &ast).unwrap();
```

(Note that using `map` and `collect` means we don't need to give our own size hint to the vector, `Chars` does it for us.)

Reply:
thanks, I like! :rocket: done! :slightly_smiling_face: